### PR TITLE
Keep orcharding

### PR DESCRIPTION
**Pull request by AloeareV**

### CODE REVIEW

File: lib/src/blaze/block_witness_data.rs [hunk 1]
```diff
@@ -106,9 +113,9 @@ impl BlockAndWitnessData {
         }
     }
 
-    pub async fn get_ctx_for_nf_at_height(
+    pub async fn get_compact_transaction_for_sapling_nullifier_at_height(
```
Comment (zancas) [inline lib/src/blaze/block_witness_data.rs:116]: I am way +1 on this!!!

File: lib/src/blaze/block_witness_data.rs [hunk 2]
```diff
@@ -3,10 +3,11 @@ use crate::{
     grpc_connector::GrpcConnector,
     lightclient::checkpoints::get_all_main_checkpoints,
     wallet::{
-        data::{BlockData, WalletTx, WitnessCache},
+        data::{BlockData, FromCommitment, WalletNullifier, WalletTx, WitnessCache},
```
Comment (zancas) [inline lib/src/blaze/block_witness_data.rs:6]: `WalletNullifier` is interesting terminology....

File: cli/src/main.rs [hunk 3]
```diff
@@ -238,7 +235,7 @@ pub fn main() {
         unflagged_lwd_log.push_str(lwd_logs.to_str().expect("error making lwd_datadir"));
         unflagged_lwd_log.push_str("lwd.log");
 
-        let lwd_command = Command::new(lwd_bin)
+        let _lwd_command = Command::new(lwd_bin)
```
Comment (zancas) [inline cli/src/main.rs:238]: This variable should be removed, it's being bound but the invocation is for the side-effect.

File: config/src/lib.rs [hunk 4]
```diff
@@ -316,13 +316,8 @@ impl Parameters for Network {
         match self {
             Mainnet => MAIN_NETWORK.activation_height(nu),
             Testnet => TEST_NETWORK.activation_height(nu),
-            _ => {
```
Comment (zancas) [inline config/src/lib.rs:319]: This comment went obsolete when nu5 activated?

File: cli/src/main.rs (hunk 3 above)
Comment (AloeareV) [reply cli/src/main.rs:238]: Wasn't sure, so I erred on the side of not changing the semantics. Will do

File: lib/src/blaze/block_witness_data.rs [hunk 5]
```diff
@@ -21,9 +23,9 @@ use tokio::{
     time::sleep,
 };
 use zcash_primitives::{
-    consensus::BlockHeight,
-    merkle_tree::{CommitmentTree, IncrementalWitness},
-    sapling::{Node, Nullifier},
```
Comment (zancas) [inline lib/src/blaze/block_witness_data.rs:26]: Now inside a variant of `WalletNullifier`?

File: lib/src/blaze/block_witness_data.rs (hunk 2 above)
Comment (AloeareV) [reply lib/src/blaze/block_witness_data.rs:6]: Keeping WalletNullifier (at least for now) in order to avoid adding separate channels for sapling/orchard nullifiers.

File: lib/src/wallet/transactions.rs [hunk 6]
```diff
@@ -213,33 +219,52 @@ impl WalletTxns {
...
-                    } else {
-                        None
-                    }
-                })
+                wtx.sapling_notes
+                    .iter()
+                    .filter_map(move |snd| {
+                        if wtx.block <= before_block
+                            && snd.have_spending_key
+                            && snd.witnesses.len() > 0
```
Comment (zancas) [inline lib/src/wallet/transactions.rs:235]: Why is this necessary?

Comment (zancas) [reply lib/src/wallet/transactions.rs:235]: Can we make `snd` and `ond` more explicit?
Are they `$DOMAIN_note_descriptors`?

Comment (AloeareV) [reply lib/src/wallet/transactions.rs:235]: > Can we make `snd` and `ond` more explicit? Are they `$DOMAIN_note_descriptors`?

They are, I can change the names, yes.

File: lib/src/blaze/block_witness_data.rs [hunk 7]
```diff
@@ -21,9 +23,9 @@ use tokio::{
...
+    consensus::{BlockHeight, NetworkUpgrade, Parameters},
+    merkle_tree::{CommitmentTree, Hashable, IncrementalWitness},
+    sapling::Node as SaplingNode,
```
Comment (zancas) [inline lib/src/blaze/block_witness_data.rs:28]: A `SaplingNode` is analogous to a `MerkleHashOrchard` in this context.

File: lib/src/blaze/block_witness_data.rs [hunk 8]
```diff
@@ -457,25 +494,37 @@ impl BlockAndWitnessData {
         }
     }
 
-    pub async fn get_note_witness(
+    async fn get_note_witnesses<D, Spend, TreeGetter, OutputsFromTransaction, Node>(
```
Comment (zancas) [inline lib/src/blaze/block_witness_data.rs:497]: Is this name change correct?   `witness` to `witnesses`?

Comment (AloeareV) [reply lib/src/blaze/block_witness_data.rs]: Fixed

File: lib/src/wallet/transactions.rs [hunk 9]
```diff
@@ -213,33 +219,60 @@ impl WalletTxns {
         &self.last_txid
     }
 
-    pub fn get_notes_for_updating(&self, before_block: u64) -> Vec<(TxId, SaplingNullifier)> {
+    pub fn get_notes_for_updating(&self, before_block: u64) -> Vec<(TxId, WalletNullifier)> {
```
Comment (zancas) [inline lib/src/wallet/transactions.rs:222]: See tracking issue for non-blocking notes:   https://github.com/zingolabs/zingolib/issues/27

File: lib/src/wallet/transactions.rs [hunk 10]
```diff
@@ -332,22 +394,38 @@ impl WalletTxns {
...
+            WalletNullifier::Sapling(nf) => {
+                let mut note_data = self
+                    .current
+                    .get_mut(&txid)
+                    .unwrap()
+                    .sapling_notes
+                    .iter_mut()
+                    .find(|n| n.nullifier == *nf)
+                    .unwrap();
+                note_data.spent = Some((spent_txid.clone(), spent_at_height.into()));
```
Comment (zancas) [inline lib/src/wallet/transactions.rs:411]: `note_data.spent` should be a three variant enum

```Rust
enum Spent {  Confirmed, Unconfirmed, None }
```
`note_data.unspent` should be removed.

File: lib/src/wallet/transactions.rs [hunk 11]
```diff
@@ -399,50 +477,53 @@ impl WalletTxns {
...
-            let wtx = self.get_or_create_tx(
-                &txid,
-                BlockHeight::from(height),
-                unconfirmed,
-                timestamp as u64,
-            );
 
-            // Mark the height correctly, in case this was previously a mempool or unconfirmed tx.
-            wtx.block = height;
+        let wtx = self.get_or_create_tx(
```
Comment (zancas) [inline lib/src/wallet/transactions.rs:486]: This fn should be `get_or_create_transaction_meta_data`, or something like that.

Comment (zancas) [reply lib/src/wallet/transactions.rs:486]: `wtx` should be something like:  `spent_txn_metadata` or `spent_txn_data`

File: lib/src/wallet/transactions.rs [hunk 12]
```diff
@@ -399,50 +477,53 @@ impl WalletTxns {
...
-                .iter_mut()
-                .find(|n| n.nullifier == nullifier)
-                .map(|nd| {
-                    // Record the spent height
-                    nd.spent = Some((txid, height.into()));
-                });
+                // Since this Txid has spent some funds, output notes in this Tx that are sent to us are actually change.
+                self.check_notes_mark_change(&txid);
+
+                // Mark the source note's nullifier as spent
```
Comment (zancas) [inline lib/src/wallet/transactions.rs:510]: I agree with removing `nullifier` from this comment.

File: lib/src/wallet/transactions.rs [hunk 13]
```diff
@@ -578,50 +674,123 @@ impl WalletTxns {
...
 
-        let wtx = self.get_or_create_tx(&txid, BlockHeight::from(height), unconfirmed, timestamp);
+        let wtx = self.get_or_create_transaction_metadata(
+            &txid,
+            BlockHeight::from(height),
+            unconfirmed,
+            timestamp,
+        );
         // Update the block height, in case this was a mempool or unconfirmed tx.
         wtx.block = height;
```
Comment (zancas) [inline lib/src/wallet/transactions.rs:755]: Uhhhmm...     how is this OK?

Comment (zancas) [reply lib/src/wallet/transactions.rs:755]: Ahh...    `wtx`'s value is a `&mut`

File: lib/src/wallet/data.rs [hunk 14]
```diff
@@ -189,6 +196,58 @@ impl WitnessCache {
     //     return hex::encode(buf);
     // }
 }
+pub(crate) trait FromCommitment {
+    fn from_commitment(from: &[u8; 32]) -> Self;
+}
+
+pub(crate) trait ToBytes<const N: usize> {
```
Comment (zancas) [inline lib/src/wallet/data.rs:203]: Why not a `u8`?   Or `u16`?

Comment (zancas) [reply lib/src/wallet/data.rs:203]: The largest value of `N` I have noticed thus far is 52.

File: lib/src/wallet/data.rs [hunk 15]
```diff
@@ -189,6 +196,58 @@ impl WitnessCache {
     //     return hex::encode(buf);
     // }
 }
+pub(crate) trait FromCommitment {
```
Comment (zancas) [inline lib/src/wallet/data.rs:199]: Maybe `FromBytes`?

File: lib/src/wallet/data.rs [hunk 16]
```diff
@@ -189,6 +196,58 @@ impl WitnessCache {
...
+}
+impl FromCommitment for MerkleHashOrchard {
+    fn from_commitment(from: &[u8; 32]) -> Self {
+        Self::from_bytes(from).unwrap()
+    }
+}
+
+pub(crate) trait NoteData {
+    type Fvk: Clone;
+    type Diversifier;
```
Comment (zancas) [inline lib/src/wallet/data.rs:232]: `Diversifier` and `Note` are associated types of the `NoteData` trait (which is implemented on `OrchardNoteData` and `SaplingNoteData`).

But `struct SaplingNoteData` has fields `pub note` and `pub diversifier`.
`struct OrchardNoteData` has fields `diversifier` and `note` as well.

This smells suspicious to me..   can we not DRY these fields somehow.

Also, given the adjacent declarations and implementations...   should we simplify the `struct` names to `OrchardNote` and `SaplingNote`?

File: lib/src/wallet/data.rs [hunk 17]
```diff
@@ -189,6 +196,58 @@ impl WitnessCache {
...
+        note: Self::Note,
+        witnesses: WitnessCache<Self::Node>,
+        nullifier: Self::Nullifier,
+        spent: Option<(TxId, u32)>,
+        unconfirmed_spent: Option<(TxId, u32)>,
+        memo: Option<Memo>,
+        is_change: bool,
+        have_spending_key: bool,
+    ) -> Self;
+    fn nullifier(&self) -> Self::Nullifier;
```
Comment (zancas) [inline lib/src/wallet/data.rs:248]: How about `get_nullifier` and `get_witnesses`?

File: lib/src/wallet/data.rs (hunk 16 above)
Comment (zancas) [reply lib/src/wallet/data.rs:232]: I gave myself a non-blocking issue to follow up on this:

https://github.com/zingolabs/zingolib/issues/57

**Review by zancas — APPROVED** [review]

File: lib/src/wallet/transactions.rs (hunk 10 above)
Comment (zancas) [reply lib/src/wallet/transactions.rs:411]: Non-blocking, tracking with: #54

File: lib/src/wallet/data.rs (hunk 14 above)
Comment (zancas) [reply lib/src/wallet/data.rs:203]: Non-blocking.

File: lib/src/wallet/data.rs (hunk 15 above)
Comment (zancas) [reply lib/src/wallet/data.rs:199]: Addressed here:   https://github.com/AloeareV/zingolib/pull/3

**Review by zancas — APPROVED** [review]

**Review by zancas — APPROVED** [review]
